// typescript/dist/plugin.cjs
Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });
const require_transform = require('./transform-CosQW1lp.cjs');
let unplugin = require("unplugin");

//#region src/plugin/index.ts
/**
* Unplugin for TypeScript operator overloading
*/
/**
* Factory function for creating the tsover unplugin
*/
const unpluginFactory = (options = {}) => {
	const { tsconfigPath, include, exclude = "node_modules/**" } = options;
	let programManager;
	const includePatterns = include ? Array.isArray(include) ? include : [include] : [
		"**/*.ts",
		"**/*.tsx",
		"**/*.mts",
		"**/*.cts"
	];
	const excludePatterns = exclude ? Array.isArray(exclude) ? exclude : [exclude] : ["node_modules/**"];
	return {
		name: "tsover",
		enforce: "pre",
		buildStart() {
			try {
				programManager = new require_transform.ProgramManager({ tsconfigPath });
			} catch (error) {
				if (error instanceof Error) throw new Error(`[tsover] Failed to initialize: ${error.message}`, { cause: error });
				throw error;
			}
		},
		transformInclude(id) {
			return shouldTransform(id, includePatterns, excludePatterns);
		},
		transform(code, id) {
			try {
				programManager.refresh();
				const sourceFile = programManager.getSourceFile(id);
				if (!sourceFile) return null;
				const result = require_transform.transformSourceFile({
					checker: programManager.getTypeChecker(),
					sourceFile
				});
				if (result.code === code) return null;
				return {
					code: result.code,
					map: result.map
				};
			} catch (error) {
				if (error instanceof Error) throw new Error(`[tsover] Failed to transform ${id}:\n${error.message}`, { cause: error });
				throw error;
			}
		},
		buildEnd() {
			if (programManager) programManager.destroy();
		}
	};
};
/**
* Create the unplugin instance
*/
const unplugin$1 = (0, unplugin.createUnplugin)(unpluginFactory);
/**
* Vite plugin - for backward compatibility and convenience
* @deprecated Use `unplugin.vite` instead
*/
const vitePlugin = unplugin$1.vite;
/**
* Webpack plugin
*/
const webpackPlugin = unplugin$1.webpack;
/**
* Rollup plugin
*/
const rollupPlugin = unplugin$1.rollup;
/**
* esbuild plugin
*/
const esbuildPlugin = unplugin$1.esbuild;
/**
* Rspack plugin
*/
const rspackPlugin = unplugin$1.rspack;
/**
* Check if a file should be transformed based on include/exclude patterns
*/
function shouldTransform(id, include, exclude) {
	const { minimatch } = require("minimatch");
	for (const pattern of exclude) if (minimatch(id, pattern) || minimatch(id, pattern, { matchBase: true })) return false;
	for (const pattern of include) if (minimatch(id, pattern) || minimatch(id, pattern, { matchBase: true })) return true;
	return false;
}

//#endregion
exports.esbuildPlugin = esbuildPlugin;
exports.rollupPlugin = rollupPlugin;
exports.rspackPlugin = rspackPlugin;
exports.unplugin = unplugin$1;
exports.unpluginFactory = unpluginFactory;
exports.vitePlugin = vitePlugin;
exports.webpackPlugin = webpackPlugin;
