// typescript/dist/transform-CosQW1lp.cjs
//#region \0rolldown/runtime.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __copyProps = (to, from, except, desc) => {
	if (from && typeof from === "object" || typeof from === "function") {
		for (var keys = __getOwnPropNames(from), i = 0, n = keys.length, key; i < n; i++) {
			key = keys[i];
			if (!__hasOwnProp.call(to, key) && key !== except) {
				__defProp(to, key, {
					get: ((k) => from[k]).bind(null, key),
					enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
				});
			}
		}
	}
	return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", {
	value: mod,
	enumerable: true
}) : target, mod));

//#endregion
let tsover = require("tsover");
tsover = __toESM(tsover);
let path = require("path");
path = __toESM(path);

//#region src/plugin/type-checker.ts
/**
* TypeScript program manager with caching
*/
var ProgramManager = class {
	constructor(options = {}) {
		this.program = null;
		this.configPath = this.resolveConfigPath(options.tsconfigPath);
		this.host = tsover.default.createCompilerHost({});
		this.options = this.loadCompilerOptions();
		this.program = this.createProgram();
	}
	resolveConfigPath(tsconfigPath) {
		if (tsconfigPath) return path.resolve(tsconfigPath);
		const configPath = tsover.default.findConfigFile(process.cwd(), tsover.default.sys.fileExists, "tsconfig.json");
		if (!configPath) throw new Error("Could not find tsconfig.json. Please ensure you have a tsconfig.json in your project root or specify a custom path via the tsconfigPath option.");
		return configPath;
	}
	loadCompilerOptions() {
		const configFile = tsover.default.readConfigFile(this.configPath, tsover.default.sys.readFile);
		if (configFile.error) {
			const message = tsover.default.flattenDiagnosticMessageText(configFile.error.messageText, "\n");
			throw new Error(`Failed to read tsconfig.json at ${this.configPath}:\n${message}`);
		}
		const parsedConfig = tsover.default.parseJsonConfigFileContent(configFile.config, tsover.default.sys, path.dirname(this.configPath));
		if (parsedConfig.errors.length > 0) {
			const errors = parsedConfig.errors.map((d) => tsover.default.flattenDiagnosticMessageText(d.messageText, "\n")).join("\n");
			throw new Error(`Failed to parse tsconfig.json at ${this.configPath}:\n${errors}`);
		}
		return parsedConfig.options;
	}
	createProgram() {
		const configFile = tsover.default.readConfigFile(this.configPath, tsover.default.sys.readFile);
		const parsedConfig = tsover.default.parseJsonConfigFileContent(configFile.config, tsover.default.sys, path.dirname(this.configPath));
		const program = tsover.default.createProgram(parsedConfig.fileNames, this.options, this.host);
		const errors = tsover.default.getPreEmitDiagnostics(program).filter((d) => d.category === tsover.default.DiagnosticCategory.Error);
		if (errors.length > 0) {
			const errorMessages = errors.map((diagnostic) => {
				const message = tsover.default.flattenDiagnosticMessageText(diagnostic.messageText, "\n");
				if (diagnostic.file && diagnostic.start !== void 0) {
					const { line, character } = diagnostic.file.getLineAndCharacterOfPosition(diagnostic.start);
					return `  at ${diagnostic.file.fileName}:${line + 1}:${character + 1}\n    ${message}`;
				}
				return `  ${message}`;
			}).join("\n\n");
			console.warn(`[tsover] Type checking warnings:\n\n${errorMessages}`);
		}
		return program;
	}
	getProgram() {
		if (!this.program) throw new Error("Program has been destroyed");
		return this.program;
	}
	getTypeChecker() {
		return this.getProgram().getTypeChecker();
	}
	getSourceFile(fileName) {
		return this.getProgram().getSourceFile(fileName);
	}
	refresh() {
		this.program = this.createProgram();
	}
	destroy() {
		this.program = null;
	}
};

//#endregion
//#region src/plugin/type-detection.ts
/**
* Check if either operand of a binary expression has operator overloading
*/
function shouldTransformBinaryExpression(node, checker) {
	if (checker.__tsover__isInUseGpuScope(node)) return false;
	return checker.__tsover__couldHaveOverloadedOperators(node.left, node.operatorToken.kind, node.right, checker.getTypeAtLocation(node.left), checker.getTypeAtLocation(node.right));
}

//#endregion
//#region src/plugin/utils.ts
/**
* Utility functions for the plugin
*/
const opToRuntimeFn = {
	[tsover.default.SyntaxKind.PlusToken]: "add",
	[tsover.default.SyntaxKind.PlusEqualsToken]: "add",
	[tsover.default.SyntaxKind.MinusToken]: "sub",
	[tsover.default.SyntaxKind.MinusEqualsToken]: "sub",
	[tsover.default.SyntaxKind.AsteriskToken]: "mul",
	[tsover.default.SyntaxKind.AsteriskEqualsToken]: "mul",
	[tsover.default.SyntaxKind.SlashToken]: "div",
	[tsover.default.SyntaxKind.SlashEqualsToken]: "div",
	[tsover.default.SyntaxKind.AsteriskAsteriskToken]: "pow",
	[tsover.default.SyntaxKind.AsteriskAsteriskEqualsToken]: "pow",
	[tsover.default.SyntaxKind.PercentToken]: "mod",
	[tsover.default.SyntaxKind.PercentEqualsToken]: "mod"
};
const assignmentOps = [
	tsover.default.SyntaxKind.PlusEqualsToken,
	tsover.default.SyntaxKind.MinusEqualsToken,
	tsover.default.SyntaxKind.AsteriskEqualsToken,
	tsover.default.SyntaxKind.SlashEqualsToken,
	tsover.default.SyntaxKind.AsteriskAsteriskEqualsToken,
	tsover.default.SyntaxKind.PercentEqualsToken
];

//#endregion
//#region src/plugin/transform.ts
/**
* AST transformation logic
*/
/**
* Transform a source file to replace + operators with __tsover_add() calls
*/
function transformSourceFile(options) {
	const { checker, sourceFile } = options;
	const transformer = (context) => {
		return (sourceFile) => {
			function visit(node) {
				if (tsover.default.isBinaryExpression(node)) {
					const runtimeFn = opToRuntimeFn[node.operatorToken.kind];
					if (shouldTransformBinaryExpression(node, checker) && runtimeFn) {
						const left = tsover.default.visitNode(node.left, visit);
						const right = tsover.default.visitNode(node.right, visit);
						const callExpression = tsover.default.factory.createCallExpression(tsover.default.factory.createIdentifier(`__tsover_${runtimeFn}`), void 0, [left, right]);
						if (assignmentOps.includes(node.operatorToken.kind)) return tsover.default.factory.createAssignment(left, callExpression);
						else return callExpression;
					}
				}
				return tsover.default.visitEachChild(node, visit, context);
			}
			return tsover.default.visitNode(sourceFile, visit);
		};
	};
	const result = tsover.default.transform(sourceFile, [transformer]);
	const transformedSourceFile = result.transformed[0];
	const transformedCode = tsover.default.createPrinter({ newLine: tsover.default.NewLineKind.LineFeed }).printFile(transformedSourceFile);
	result.dispose();
	return { code: transformedCode };
}

//#endregion
Object.defineProperty(exports, 'ProgramManager', {
  enumerable: true,
  get: function () {
    return ProgramManager;
  }
});
Object.defineProperty(exports, '__toESM', {
  enumerable: true,
  get: function () {
    return __toESM;
  }
});
Object.defineProperty(exports, 'transformSourceFile', {
  enumerable: true,
  get: function () {
    return transformSourceFile;
  }
});
